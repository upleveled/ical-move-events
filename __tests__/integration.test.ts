import { readFileSync, rmSync, writeFileSync } from 'node:fs';
import { dirname, relative } from 'node:path';
import { fileURLToPath } from 'node:url';
import { execaCommandSync } from 'execa';
import { expect, test } from 'vitest';

const testsDir = relative(
  process.cwd(),
  dirname(fileURLToPath(import.meta.url)),
);
const inputIcsFilePath = `${testsDir}/ical-move-events-input-calendar.ics`;
const outputIcsFilePath = inputIcsFilePath.replace('.ics', '-moved.ics');
rmSync(outputIcsFilePath, { force: true });

function getNormalizedOutputFileContents() {
  return readFileSync(outputIcsFilePath, 'utf-8')
    .replaceAll(/\nUID:[a-z0-9-]+/g, '')
    .replaceAll(/(\nDTSTAMP:).+/g, '$1');
}

// Set timezone on GitHub Actions
if (process.platform === 'linux') {
  execaCommandSync('sudo timedatectl set-timezone Europe/Vienna');
}

test('moves calendar entries and saves file', () => {
  const { stdout } = execaCommandSync(
    `pnpm --silent dev ${inputIcsFilePath} --start 2021-08-23 --end 2021-09-03`,
  );
  expect(stdout).toMatchSnapshot();

  const outputFileContents = getNormalizedOutputFileContents();
  expect(outputFileContents).toMatchSnapshot();
  rmSync(outputIcsFilePath);
});

test('throws error if output file location already exists', () => {
  writeFileSync(outputIcsFilePath, '');
  const { stderr } = execaCommandSync(
    `pnpm --silent dev ${inputIcsFilePath} --start 2021-08-23 --end 2021-09-03`,
    { reject: false },
  );
  expect(stderr.replace(/node:\d+/, 'node:')).toMatchSnapshot();
  rmSync(outputIcsFilePath);
});

test('moves calendar entries and saves file, taking into account holiday and timezone change from Daylight Savings Time', () => {
  const { stdout } = execaCommandSync(
    `pnpm --silent dev ${inputIcsFilePath} --start 2021-11-01 --end 2021-11-12`,
  );
  expect(stdout).toMatchSnapshot();

  const outputFileContents = getNormalizedOutputFileContents();
  expect(outputFileContents).toMatchSnapshot();
  rmSync(outputIcsFilePath);
});
