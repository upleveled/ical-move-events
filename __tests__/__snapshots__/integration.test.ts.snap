// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moves calendar entries and saves file 1`] = `
"$ node --loader ts-node/esm/transpile-only.mjs index.ts __tests__/ical-move-events-input-calendar.ics 2021-08-23
Calendar events updated and saved in __tests__/ical-move-events-input-calendar-moved.ics"
`;

exports[`moves calendar entries and saves file 2`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//sebbo.net//ical-generator//EN
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210831T070000Z
DTEND:20210831T160000Z
SUMMARY:üéâ Holiday
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210824T103000Z
DTEND:20210824T160000Z
SUMMARY:Project Time 2
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210823T103000Z
DTEND:20210823T160000Z
SUMMARY:Project Time 1
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210825T103000Z
DTEND:20210825T160000Z
SUMMARY:Project Time 3
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210824T073000Z
DTEND:20210824T103000Z
SUMMARY:üßë‚Äçüè´ Lecture 2
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210825T073000Z
DTEND:20210825T103000Z
SUMMARY:üßë‚Äçüè´ Lecture 3
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210826T073000Z
DTEND:20210826T103000Z
SUMMARY:üßë‚Äçüè´ Lecture 4
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210823T220000Z
DTEND:20210826T220000Z
SUMMARY:üèó Project 3 (stretch) (3 days)
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210822T220000Z
DTEND:20210825T220000Z
SUMMARY:üèó Project 1 (3 days)
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210825T220000Z
DTEND:20210903T220000Z
SUMMARY:üèó Project 2 (6 days)
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210824T070000Z
DTEND:20210824T073000Z
RRULE:FREQ=WEEKLY;INTERVAL=1;WKST=MO;COUNT=9;BYDAY=FR,MO,TH,TU,WE;BYHOUR=7
 ;BYMINUTE=0;BYSECOND=0
SUMMARY:Standup (repeating)
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210823T070000Z
DTEND:20210823T103000Z
SUMMARY:üßë‚Äçüè´ Lecture 1
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210826T103000Z
DTEND:20210826T160000Z
SUMMARY:Project Time 4
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210827T103000Z
DTEND:20210827T160000Z
SUMMARY:Project Time 5
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210827T073000Z
DTEND:20210827T103000Z
SUMMARY:üßë‚Äçüè´ Lecture 5
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210830T073000Z
DTEND:20210830T103000Z
SUMMARY:üßë‚Äçüè´ Lecture 6
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210830T103000Z
DTEND:20210830T160000Z
SUMMARY:Project Time 6
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210902T073000Z
DTEND:20210902T160000Z
SUMMARY:Project Time 8
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210901T073000Z
DTEND:20210901T160000Z
SUMMARY:Project Time 7
DESCRIPTION:
END:VEVENT
BEGIN:VEVENT

SEQUENCE:0
DTSTAMP:
DTSTART:20210903T073000Z
DTEND:20210903T160000Z
SUMMARY:Project Time 9
DESCRIPTION:
END:VEVENT
END:VCALENDAR"
`;

exports[`throws error if output file location already exists 1`] = `
"(node:) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use \`node --trace-warnings ...\` to show where the warning was created)
Error: Output file at __tests__/ical-move-events-input-calendar-moved.ics already exists!
error Command failed with exit code 1."
`;
